### DISCUSSION TITLE
fzf-lua always showing processing indicator

### DISCUSSION THREAD

**Discussion by HendrikThePendric** [LazyVim/LazyVim#5264] [Q&A] [answered]
After upgrading to v14, I am now using fzf-lua instead of telescope. Whenever I open the file picker (`space + f + f` / `space + f + F` / `space + f +g`) I see the following animated indicator at the end of the input:
<img width="557" alt="Screenshot 2024-12-27 at 21 23 29" src="https://github.com/user-attachments/assets/c1de8790-ea1f-4ac6-bc7f-2caf4b4ea082" />

It seems to suggest that the files are still being indexed or some other process is still happening in the background. I suspected it could be to do with the `node_modules` dir in my JavaScript projects. But this didn't seem to be the case because even after I deleted that dir, the indicator still shows up. On the other hand, when in a repo with only a few files like my dotfiles repo, I do not see this indicator.

I'm not sure if what I am seeing is purely a visual glitch or if anything is actually wrong, but the UI look a bit broken this way.

How can I solve this?

PS: Perhaps I should have been posting this in the fzf-lua repo.... Not sure if this is related to the plugin or its configuration.

**Comment by dpetka2001** [comment]
I don't see it myself. This is how it looks for my in the folder `~/.config` (my $HOME would have taken much longer)

[Screencast_2024-12-28-00-20-56.webm](https://github.com/user-attachments/assets/71d0e6c1-8306-436d-962e-114892c6bdb6)

Of course in my Neovim configuration folder it opens instantly, since the number of files is significantly smaller.

**Comment by HendrikThePendric** [comment]
Thanks for showing that @dpetka2001, good to have confirmation that the "spinner" is supposed to eventually stop.

I have noticed just now that this is very project specific, some of my projects suffer from this issue while others don't. It doesn't seem related to the number of files though because I have some larger projects which work fine. Can't put my finger on it just yet. I'll look into it some more.

**Comment by HendrikThePendric** [comment]
One interesting thing I noticed just now is that when I open NeoVim from my home dir and then open the file picker, it takes a very long time but the spinner does eventually stop once all the files have been traversed. This is interesting to me because these project dirs that suffer from this issue actually reside in my home dir 🙃

**Comment by HendrikThePendric** [comment]
No closer to a solution yet, but I can now say that this is somehow related to the `.gitignore` file. In a project that keeps showing this spinner, when I press `alt-i` (toggle-ignore) I see the file count grow and then the spinner stops. This also explains why the problem does not occur when I open NeoVim from my home dir, since that doesn't contain a `.gitginore`.

**Reply by dpetka2001** [reply]
Maybe you'd like to ask the fzf-lua maintainer for his opinion as i'm just a user and not familiar with its codebase.

**Reply by HendrikThePendric** [reply]
Thanks, I'll probably do that in the new year. I'd like to be able to get fzf-lua to produce some sort of verbose log so I can see what's actually happening. It's all a bit of a black box to me right now...

**Reply by dpetka2001** [reply]
If I recall correctly you can do `:FzfLua files debug=true` and that will give you some kind of debug information. But like I said I'm not familiar with the internals, so if you show that output to the maintainer himself I'm sure he would be able to tell you what might be wrong.

**Reply by HendrikThePendric** [reply]
This was somewhat useful. The indicator I have been talking about is referred to as a "spinner" and `fzf-lua` only tells `fzf` which color to render this in. The debug output shows which command `fzf-lua` sends to `fzf`:
```
fzf --info 'inline-right' --highlight-line --print-query --layout 'reverse' --height '100%' --ansi --prompt ' ' --color 'separator:#89b4fb,fg:#cdd6f5,header:#89b4fb,info:#6c7087,scrollbar:#89b4fb,query:#cdd6f5:regular,pointer:#f5c2e8,marker:#f5c2e8,gutter:#1e1e2f,hl+:#89b4fb,bg+:#2a2b3d,hl:#89b4fb,prompt:#89b4fb,fg+:#cdd6f5,spinner:#f5c2e8,border:#89b4fb,bg:#1e1e2f' --border 'none'
```
_(There is more but I'll omit that)_

I would have expected that `cd`ing into my project dir and issuing the command above would allow me to reproduce the issue directly in my terminal. But although I do see the exact same "search bar" as I do in NeoVim, the spinner in this case is actually not showing.

So at the very least I have some more leads to look into based on this. 

_([Here's](https://github.com/junegunn/fzf/blob/fe3a9c603e924425f80b87d76adb3e80d2bc668c/src/terminal.go#L691-L696) the code from the `fzf` repo that renders the spinner)_

**Reply by dpetka2001** [reply]
Like I said, I'm unfortunately clueless to fzf-lua codebase internals, so it'd be best if you brought this to the maintainer's attention, so that he can provide more help.

**Reply by ibhagwan** [reply]
@HendrikThePendric, unless new files are being enumerated the spinner should stop, if this never stops perhaps there’s a bug in the headless neovim wrapper process code which makes it hang, might be related to some files you got in that project that are usually just gitignore’d.

Plesse run `:FzfLua profiles` and select `fzf-native` or `max-perf`, both should pipe the fd/rg command directly to fzf without the lua wrapper, if this solves it please open a bug report in the fzf-lua repo.

There’s another slight chance that you have some remote mounted folders that are usually ignored and are very slow to enumerate (when you disable the gitignore).

For debugging you can also run `:FzfLua files multiprocess=false`, this will have bad perf as it’s all running on the main thread but if there’s an exception you’ll be able to see it in `:messages` which may give us some insight what’s  causing the issue.

**Reply by ibhagwan** [reply]
@HendrikThePendric, have you tried any of my suggestions above? It would be very helpful if you have a specific repo wheee this happens that I can git clone and reproduce this myself.

**Reply by HendrikThePendric** [reply]
Sorry @ibhagwan I've been away from my computer for a few days so have been unable to reply. Here's an example of a project where I get this problem: https://github.com/dhis2/dashboard-app

Not sure if this is relevant but I am on MacOS with M1 Max.

I'll also try your suggestions and report back.

**Reply by ibhagwan** [reply]
> Sorry @ibhagwan I've been away from my computer for a few days so have been unable to reply. Here's an example of a project where I get this problem: https://github.com/dhis2/dashboard-app
> 
> Not sure if this is relevant but I am on MacOS with M1 Max.
> 
> I'll also try your suggestions and report back.

Ty @HendrikThePendric but as you suspected the issue comes from files in your `.gitignore`, without building or running the project my total file count is 493 even with disabling the gitignore:

![image](https://github.com/user-attachments/assets/b859d146-9bbf-462a-9221-04ac2f6b2a0e)

**Reply by HendrikThePendric** [reply]
I switched to the `fzf-native` profile and still see the problem. I then switched to the `max-perf` profile and to my surprise this *did* resolve things. I say to my surprise, because my config file for the `fzf-lua` plugin already looks like this:
```lua
return {
  "ibhagwan/fzf-lua",
  opts = {
    files = {
      git_icons = false,
    },
  },
}
```

**Reply by HendrikThePendric** [reply]
@dpetka2001 sorry it seems our messages were posted almost simultaneously, so I did not reply to your message but just posted my findings. 

> without building or running the project my total file count is 493 even with disabling the gitignore

Thanks for looking into this. Yes the project has 493 files. What are you implying? That this is a large number and is causing the problem?

**Reply by ibhagwan** [reply]
@HendrikThePendric, as I mentioed I suspect the issue is with the file icons or our outside processor, `fzf-native` doesn't disable the processor but `max-perf` does.

Try my other suggestion `:FzfLua files multiprocess=false`, if my assumptions are correct it would give us a lead what's breaking.

Also, with `max-perf` what's the total count of the files? if it's a small number of files that's failing to load we can identify the missing files in the list which would be very helpful to solve this.

**Reply by ibhagwan** [reply]
> Thanks for looking into this. Yes the project has 493 files. What are you implying? That this is a large number and is causing the problem?

I'm implying that when you disable the gitignore there are other files on your system causing this issue (probably from building the project) which don't exist on my system - that's why I don't have the same issue (in the screenshot from the OP I see that's correct as it shows 494 files).

**Reply by HendrikThePendric** [reply]
OK I'll report what I see...

With `max-perf` and with other profiles I now realise that I see 494 files instead and not 493, so I am 1 off it seems.... 

When I switch back to the default profile and then do `:FzfLua files multiprocess=false` I don't see the spinner lingering and I don't get any output when I do `:messages`.

**Reply by HendrikThePendric** [reply]
> I'm implying that when you disable the gitignore there are other files on your system causing this issue (probably from building the project) which don't exist on my system - that's why I don't have the same issue (in the screenshot from the OP I see that's correct as it shows 494 files).

No, I have this problem not when I disable the gitignore, I have it when I have it enabled. Everything works as expected when I disable the gitignore (assuming disabling the gitignore means "list all files").

**Reply by HendrikThePendric** [reply]
I'm also fairly sure that when I remove the project and clone it again the problem is already there for me. But I am not sure if I tried that with this particular project. I'll try that again.

**Reply by ibhagwan** [reply]
> With max-perf and with other profiles I now realise that I see 494 files instead and not 493, so I am 1 off it seems....

Can you tell what is the full project path of the missing file?

> When I switch back to the default profile and then do `:FzfLua files multiprocess=false` I don't see the spinner lingering and I don't get any output when I do `:messages.`

Does the processing indicator stop and the count is 494?

When you run the "bugged" ` :FzfLua files` (with the indicator still spinning), is the count also 494? do you see the missing file in the list?

If so that would mean nothing fails but the wrapper process doesn't terminate leading to a continous indicator.

In any event that missing file should lead us to the solution, explore the `.gitignore` as it will give some hints.

Just a hunch, could it be the mac's `.DS_Store` file?

**Reply by HendrikThePendric** [reply]
Just to confirm: the 494 VS 493 files issue is not relevant. When I remove the project and re-clone I also get 493 files. That 494th file was just an environment variables for Cypress. And more importantly with 493 files I still see that spinner:

<img width="1796" alt="Screenshot 2025-01-02 at 19 47 13" src="https://github.com/user-attachments/assets/4ec275c4-0193-43ee-bdf3-1d138cb06d86" />

**Reply by ibhagwan** [reply]
> Just to confirm: the 494 VS 493 files issue is not relevant. When I remove the project and re-clone I also get 493 files. That 494th file was just an environment variables for Cypress. And more importantly with 493 files I still see that spinner:

Ok, that’s good progress I’m going to try the same project on a Mac.

Can you test with `mini.sh` so I can rule out anything config / LazyVim related triggering this? Run the below inside the project dir and press `ctrl-p`:
```sh
sh -c "$(curl -s https://raw.githubusercontent.com/ibhagwan/fzf-lua/main/scripts/mini.sh)"
```

**Reply by ibhagwan** [reply]
Tyvm @HendrikThePendric! This does reproduce on my mac, even with minimal init, the road to solve this now should hopefully be short.

**Reply by HendrikThePendric** [reply]
Did that and see the spinner still:
<img width="1840" alt="Screenshot 2025-01-02 at 19 54 51" src="https://github.com/user-attachments/assets/63a933a5-468c-4f1d-99d8-5f54a3d0afa9" />

**Reply by HendrikThePendric** [reply]
> Tyvm @HendrikThePendric! This does reproduce on my mac, even with minimal init, the road to solve this now should hopefully be short.

No thank you @ibhagwan 🙏 ! I very much appreciate you taking the time to look into this. I hope you end up finding a proper bug and not something isolated to our projects at DHIS2.

**Reply by ibhagwan** [reply]
> > Tyvm @HendrikThePendric! This does reproduce on my mac, even with minimal init, the road to solve this now should hopefully be short.
> 
> No thank you @ibhagwan 🙏 ! I very much appreciate you taking the time to look into this. I hope you end up finding a proper bug and not something isolated to our projects at DHIS2.

Rest assured this will be solved :)

**Comment by ibhagwan** [comment, marked as answer]
@HendrikThePendric, fixed in https://github.com/ibhagwan/fzf-lua/commit/a553cfd9c83888ef22b025285deec3fb828985f9

**Reply by HendrikThePendric** [reply]
Awesome. I can confirm it's fixed after updating `fzf-lua`.
